Fix for performance penalty in deep dependencies

fgarciacorona opened the pull request:
The goal of this PR is to fix #20371 (closed) #24267 and #24330 and improve the performance of csharp runtime when dealing with deep nested dependencies of .proto files.

- The fix consists in caching the recursive calls for the search of extensions.

- Two identical datasets (nested proto files) are created automatically using Bazel: one to test it with caching enabled, and another one to test it with no caching. Note: the same dataset cannot be used because once it is loaded inside the test, it cannot be unloaded.

- The created datasets have a width of 6 and a depth of 6, enough to showcase the dependency impact and to not impact the testing time for unit tests. This could be configured.

- In the dataset there is only one proto file with a message `Example` whose descriptor is loaded, and by doing so all the dependencies (and sub dependencies) that it has. 

- A set of benchmark metrics have been created to evaluate the performance impact before and after the fix.

- The assertion inside the unit test makes sure that when using metrics the number of traversed extensions (dependencies of the dependencies) when using caching, is reduced by a factor of 1000x. 

- The second assertion also makes sure that the time used to load the descriptor when using cache is less than the one without using cache.

tonyliaoss commented:
Hello @fgarciacorona, we won't be able to start reviewing your PR until the CLA has been signed.

Also, I suspect that we won't accept the newly added test protos, but we can discuss that after we start our review process.

fgarciacorona commented:
> Hello @fgarciacorona, we won't be able to start reviewing your PR until the CLA has been signed.
> 
> Also, I suspect that we won't accept the newly added test protos, but we can discuss that after we start our review process.

@tonyliaoss CLA has been passed! I cannot modify the `wait for user action` tag

fgarciacorona commented:
@tonyliaoss Is there any additional action required from my side to have the PR marked as safe for test and proceed with the testing of the PR? Additionally, I think it is missing the `C#` label.

JasonLunn commented:
Could you rebase this PR? Github thinks it is changing almost 800 files which makes review impractical.

fgarciacorona commented:
> Could you rebase this PR? Github thinks it is changing almost 800 files which makes review impractical.
> 
> Edit, nevermind about the rebase. I see what @tonyliaoss means about the test protos, though. Is there no other way to excite the issue that this PR aims to fix?

Right, there are 3 files that have actually been changed and the rest are half of them .proto and the other half the .cs generated classes from the protos.

I could only think of dynamically generating .proto files highly nested from within the testcase (if you want to reduce the number of existing files in the PR) and then somehow try to run `protoc` to generate the corresponding classes.

fgarciacorona commented:
@JasonLunn how could I have the repo marked as "safe for tests" without having to request it for every commit?

JasonLunn commented:
> @JasonLunn how could I have the repo marked as "safe for tests" without having to request it for every commit?

Humans have to apply the tag by policy for third party contributions to make sure they're safe to run on our CI infrastructure.

JasonLunn commented:
```
Failed FileDescriptor_CreateMessageWithDeepDependencies_BuildFromByteString [263 ms]
  Error Message:
     Expected: 402
  But was:  419

  Stack Trace:
     at Google.Protobuf.Reflection.DescriptorsTest.FileDescriptor_CreateMessageWithDeepDependencies_BuildFromByteString() in D:\a\protobuf\protobuf\csharp\src\Google.Protobuf.Test\Reflection\DescriptorsTest.cs:line 59

1)    at Google.Protobuf.Reflection.DescriptorsTest.FileDescriptor_CreateMessageWithDeepDependencies_BuildFromByteString() in D:\a\protobuf\protobuf\csharp\src\Google.Protobuf.Test\Reflection\DescriptorsTest.cs:line 59
```

Interestingly, this only fails on the Windows test configuration, but it passed on Linux...

fgarciacorona commented:
> ```
> Failed FileDescriptor_CreateMessageWithDeepDependencies_BuildFromByteString [263 ms]
>   Error Message:
>      Expected: 402
>   But was:  419
> 
>   Stack Trace:
>      at Google.Protobuf.Reflection.DescriptorsTest.FileDescriptor_CreateMessageWithDeepDependencies_BuildFromByteString() in D:\a\protobuf\protobuf\csharp\src\Google.Protobuf.Test\Reflection\DescriptorsTest.cs:line 59
> 
> 1)    at Google.Protobuf.Reflection.DescriptorsTest.FileDescriptor_CreateMessageWithDeepDependencies_BuildFromByteString() in D:\a\protobuf\protobuf\csharp\src\Google.Protobuf.Test\Reflection\DescriptorsTest.cs:line 59
> ```
> 
> Interestingly, this only fails on the Windows test configuration, but it passed on Linux...

@JasonLunn the reason why it was not failing in linux and arm64 is because the metrics for benchmarking are only generated for Debug (windows) but not for Release (linux and arm64). 

Additionally we were storing the benchmark metrics in Debug mode in a static variable that was impacted by all unit tests. This needs to be reset before the specific test that looked at them.

fgarciacorona commented:
@JasonLunn would it be possible to remove the assigned reviewers and I will do a rebase and hopefully based on that the right reviewers will be assigned? I think because I didn't rebase it properly the first time a lot of commits from main triggered so many people.

fgarciacorona commented:
> It would expedite review if you could refactor the PR so that there are not ~800 files. I assume you have a script that generated the deeply nested test `.proto` files in the first place - could you use that in a `genrule` to generate the test input files at test time so that they don't have to be individually reviewed?

The script that we have anonymized our data, so the proto files are based on actual data. But thanks for the feedback, knowing that a generator could also be used, I will focus on that. 

Question though: do the generated `.cs` files need to be also committed? because I see them committed in the repo for other test `.proto` files, and that would eventually defeat the purpose (we would have ~400 only). Additionally if we don't commit the generated `.cs` files I am not sure whether it would be an accepted approach to have the test project not able to build if the test classes are not committed.

fgarciacorona committed:
Fix for issue 20371 related to deep depencies.
Add support for deep dependency testing with and without caching in proto generation

fgarciacorona committed:
Update .gitignore and generate_protos.sh to include paths for auto-generated test classes

fgarciacorona committed:
Adapted code to be compatible with bash GNU and BSD and simplified syntax

JasonLunn committed:
Merge branch 'main' into fix_20371_csharp_deep_dependencies
